File: castella/protocols/painter.py
"""Painter protocol definitions - split into required core and optional capabilities."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Protocol, runtime_checkable

if TYPE_CHECKING:
    try:
        import numpy as np
    except ImportError:
        pass

from castella.models.geometry import Circle, Point, Rect, Size
from castella.models.font import FontMetrics
from castella.models.style import Style


@runtime_checkable
class BasePainter(Protocol):
    """Core painter protocol - all painters MUST implement these."""

    def clear_all(self) -> None:
        """Clear the entire drawing surface."""
        ...

    def fill_rect(self, rect: Rect) -> None:
        """Fill a rectangle with the current fill style."""
        ...

    def stroke_rect(self, rect: Rect) -> None:
        """Stroke a rectangle outline with the current stroke style."""
        ...

    def translate(self, pos: Point) -> None:
        """Translate the canvas origin."""
        ...

    def clip(self, rect: Rect) -> None:
        """Set the clipping region."""
        ...

    def fill_text(self, text: str, pos: Point, max_width: float | None) -> None:
        """Draw filled text at the given position."""
        ...

    def stroke_text(self, text: str, pos: Point, max_width: float | None) -> None:
        """Draw stroked text at the given position."""
        ...

    def measure_text(self, text: str) -> float:
        """Measure the width of text with the current font."""
        ...

    def get_font_metrics(self) -> FontMetrics:
        """Get metrics for the current font."""
        ...

    def save(self) -> None:
        """Save the current canvas state."""
        ...

    def restore(self) -> None:
        """Restore the previously saved canvas state."""
        ...

    def style(self, style: Style) -> None:
        """Set the current drawing style."""
        ...

    def flush(self) -> None:
        """Flush pending drawing operations."""
        ...


@runtime_checkable
class CircleCapable(Protocol):
    """Optional: Circle drawing capability."""

    def fill_circle(self, circle: Circle) -> None:
        """Fill a circle with the current fill style."""
        ...

    def stroke_circle(self, circle: Circle) -> None:
        """Stroke a circle outline with the current stroke style."""
        ...


@runtime_checkable
class LocalImageCapable(Protocol):
    """Optional: Local file image capability."""

    def draw_image(self, file_path: str, rect: Rect, use_cache: bool = True) -> None:
        """Draw an image from a local file."""
        ...

    def measure_image(self, file_path: str, use_cache: bool = True) -> Size:
        """Measure the size of an image from a local file."""
        ...


@runtime_checkable
class SyncNetworkImageCapable(Protocol):
    """Optional: Synchronous network image capability."""

    def draw_net_image(self, url: str, rect: Rect, use_cache: bool = True) -> None:
        """Draw an image from a network URL (synchronously)."""
        ...

    def measure_net_image(self, url: str, use_cache: bool = True) -> Size:
        """Measure the size of an image from a network URL (synchronously)."""
        ...


@runtime_checkable
class AsyncNetworkImageCapable(Protocol):
    """Optional: Asynchronous network image capability."""

    def get_net_image_async(self, name: str, url: str, callback: Any) -> None:
        """Fetch an image from a network URL asynchronously."""
        ...

    def draw_image_object(self, img: Any, x: float, y: float) -> None:
        """Draw a pre-loaded image object."""
        ...


@runtime_checkable
class NumpyImageCapable(Protocol):
    """Optional: Numpy array as image capability."""

    def measure_np_array_as_an_image(self, array: "np.ndarray") -> Size:
        """Measure the size of a numpy array as an image."""
        ...

    def draw_np_array_as_an_image(
        self, array: "np.ndarray", x: float, y: float
    ) -> None:
        """Draw a numpy array as an image at the given position."""
        ...

    def draw_np_array_as_an_image_rect(self, array: "np.ndarray", rect: Rect) -> None:
        """Draw a numpy array as an image within a rectangle."""
        ...


@runtime_checkable
class AsyncNumpyImageCapable(Protocol):
    """Optional: Async numpy image capability (for web)."""

    def get_numpy_image_async(self, array: "np.ndarray", callback: Any) -> None:
        """Convert numpy array to image asynchronously."""
        ...

    def draw_image_object(self, img: Any, x: float, y: float) -> None:
        """Draw a pre-loaded image object."""
        ...


@runtime_checkable
class CaretDrawable(Protocol):
    """Optional: Caret drawing for text input (terminal)."""

    def draw_caret(self, pos: Point, height: int) -> None:
        """Draw a text input caret."""
        ...


# Full Painter type alias for backward compatibility
# Combines all capabilities (not all painters implement all)
Painter = BasePainter

File: castella-skia-core/src/painter.rs
//! SkiaPainter implementation - the core painting API.
//!
//! This implements the BasePainter protocol from castella.protocols.painter.

use skia_safe::{
    Canvas, ClipOp, Color, Font, FontMgr, Paint, PaintStyle, RRect, Rect,
    textlayout::{FontCollection, ParagraphBuilder, ParagraphStyle, TextStyle},
};
use std::cell::RefCell;

use crate::error::Result;
use crate::font::{create_font, get_metrics};
use crate::image::{load_image, measure_image};
use crate::surface::Surface;
use crate::types::{parse_color, FontMetrics, Style};

// Thread-local FontCollection cache (FontCollection is not Send/Sync)
thread_local! {
    static FONT_COLLECTION: RefCell<Option<FontCollection>> = const { RefCell::new(None) };
}

fn get_font_collection() -> FontCollection {
    FONT_COLLECTION.with(|cell| {
        let mut cache = cell.borrow_mut();
        if cache.is_none() {
            let mut fc = FontCollection::new();
            fc.set_default_font_manager(FontMgr::default(), None);
            *cache = Some(fc);
        }
        cache.as_ref().unwrap().clone()
    })
}

/// State saved by save() and restored by restore().
#[derive(Clone)]
struct PainterState {
    style: Style,
    font: Font,
}

/// SkiaPainter - implements BasePainter protocol.
///
/// Provides GPU-accelerated 2D drawing using Skia.
/// Note: SkiaPainter is not Send-safe because it holds a reference to Surface.
pub struct SkiaPainter<'a> {
    surface: &'a mut Surface,
    current_style: Style,
    current_font: Font,
    state_stack: Vec<PainterState>,
}

impl<'a> SkiaPainter<'a> {
    /// Create a new SkiaPainter from a Surface.
    pub fn new(surface: &'a mut Surface) -> Self {
        Self {
            surface,
            current_style: Style::default(),
            current_font: create_font(None, 14.0),
            state_stack: Vec::new(),
        }
    }

    // ========== BasePainter Required Methods ==========

    /// Clear the entire drawing surface with background color.
    pub fn clear_all(&mut self) {
        let color = self.current_style.fill_color
            .as_ref()
            .map(|c| parse_color(c))
            .unwrap_or(Color::WHITE);

        self.with_canvas(|canvas| {
            canvas.clear(color);
        });
    }

    /// Fill a rectangle with the current fill style.
    pub fn fill_rect(&mut self, x: f32, y: f32, width: f32, height: f32) {
        let radius = self.current_style.border_radius;

        // Draw shadow first (behind the shape)
        // Clone shadow to avoid borrow checker issues
        if let Some(shadow) = self.current_style.shadow.clone() {
            self.draw_shadow(x, y, width, height, &shadow, radius);
        }

        let rect = Rect::from_xywh(x, y, width, height);
        let paint = self.make_fill_paint();

        self.with_canvas(|canvas| {
            if radius > 0.0 {
                let rrect = RRect::new_rect_xy(rect, radius, radius);
                canvas.draw_rrect(rrect, &paint);
            } else {
                canvas.draw_rect(rect, &paint);
            }
        });
    }

    /// Draw a drop shadow behind a rectangle.
    fn draw_shadow(
        &mut self,
        x: f32,
        y: f32,
        width: f32,
        height: f32,
        shadow: &crate::types::Shadow,
        border_radius: f32,
    ) {
        let mut paint = Paint::default();
        paint.set_style(PaintStyle::Fill);
        paint.set_anti_alias(true);
        paint.set_color(parse_color(&shadow.color));

        if let Some(filter) = shadow.make_blur_filter() {
            paint.set_mask_filter(filter);
        }

        // Offset rectangle for shadow
        let shadow_rect = Rect::from_xywh(
            x + shadow.offset_x,
            y + shadow.offset_y,
            width,
            height,
        );

        self.with_canvas(|canvas| {
            if border_radius > 0.0 {
                let rrect = RRect::new_rect_xy(shadow_rect, border_radius, border_radius);
                canvas.draw_rrect(rrect, &paint);
            } else {
                canvas.draw_rect(shadow_rect, &paint);
            }
        });
    }

    /// Stroke a rectangle outline with the current stroke style.
    pub fn stroke_rect(&mut self, x: f32, y: f32, width: f32, height: f32) {
        let rect = Rect::from_xywh(x, y, width, height);
        let paint = self.make_stroke_paint();
        let radius = self.current_style.border_radius;

        self.with_canvas(|canvas| {
            if radius > 0.0 {
                let rrect = RRect::new_rect_xy(rect, radius, radius);
                canvas.draw_rrect(rrect, &paint);
            } else {
                canvas.draw_rect(rect, &paint);
            }
        });
    }

    /// Translate the canvas origin.
    pub fn translate(&mut self, x: f32, y: f32) {
        self.with_canvas(|canvas| {
            canvas.translate((x, y));
        });
    }

    /// Scale the canvas.
    pub fn scale(&mut self, sx: f32, sy: f32) {
        self.with_canvas(|canvas| {
            canvas.scale((sx, sy));
        });
    }

    /// Set the clipping region.
    pub fn clip(&mut self, x: f32, y: f32, width: f32, height: f32) {
        let rect = Rect::from_xywh(x, y, width, height);
        self.with_canvas(|canvas| {
            canvas.clip_rect(rect, ClipOp::Intersect, true);
        });
    }

    /// Draw filled text at the given position.
    pub fn fill_text(&mut self, text: &str, x: f32, y: f32, max_width: Option<f32>) {
        let paint = self.make_fill_paint();
        self.draw_text_internal(text, x, y, max_width, &paint);
    }

    /// Draw stroked text at the given position.
    pub fn stroke_text(&mut self, text: &str, x: f32, y: f32, max_width: Option<f32>) {
        let paint = self.make_stroke_paint();
        self.draw_text_internal(text, x, y, max_width, &paint);
    }

    /// Measure the width of text with the current font (with fallback support).
    pub fn measure_text(&self, text: &str) -> f32 {
        if text.is_empty() {
            return 0.0;
        }

        let font_collection = get_font_collection();
        let paragraph_style = ParagraphStyle::new();

        let mut text_style = TextStyle::new();
        text_style.set_font_size(self.current_font.size());
        if let Some(ref family) = self.current_style.font_family {
            text_style.set_font_families(&[family.as_str()]);
        }

        let mut builder = ParagraphBuilder::new(&paragraph_style, font_collection);
        builder.push_style(&text_style);
        builder.add_text(text);

        let mut paragraph = builder.build();
        paragraph.layout(f32::MAX);

        paragraph.max_intrinsic_width()
    }

    /// Get metrics for the current font.
    pub fn get_font_metrics(&self) -> FontMetrics {
        get_metrics(&self.current_font)
    }

    /// Save the current canvas state.
    pub fn save(&mut self) {
        // Save style state
        self.state_stack.push(PainterState {
            style: self.current_style.clone(),
            font: self.current_font.clone(),
        });

        // Save canvas state
        self.with_canvas(|canvas| {
            canvas.save();
        });
    }

    /// Restore the previously saved canvas state.
    pub fn restore(&mut self) {
        // Restore style state
        if let Some(state) = self.state_stack.pop() {
            self.current_style = state.style;
            self.current_font = state.font;
        }

        // Restore canvas state
        self.with_canvas(|canvas| {
            canvas.restore();
        });
    }

    /// Set the current drawing style.
    pub fn set_style(&mut self, style: &Style) {
        // Only recreate font if font settings actually changed
        let font_changed = style.font_family != self.current_style.font_family
            || (style.font_size - self.current_style.font_size).abs() > 0.001;

        self.current_style = style.clone();

        if font_changed {
            let family = style.font_family.as_deref();
            let size = style.font_size;
            self.current_font = create_font(family, size);
        }
    }

    /// Get the current drawing style.
    pub fn get_style(&self) -> &Style {
        &self.current_style
    }

    /// Flush pending drawing operations.
    pub fn flush(&mut self) {
        // Note: actual flushing happens when swapping buffers
        // The surface's flush_and_submit is called separately
    }

    // ========== CircleCapable Methods ==========

    /// Fill a circle with the current fill style.
    pub fn fill_circle(&mut self, cx: f32, cy: f32, radius: f32) {
        let paint = self.make_fill_paint();
        self.with_canvas(|canvas| {
            canvas.draw_circle((cx, cy), radius, &paint);
        });
    }

    /// Stroke a circle outline with the current stroke style.
    pub fn stroke_circle(&mut self, cx: f32, cy: f32, radius: f32) {
        let paint = self.make_stroke_paint();
        self.with_canvas(|canvas| {
            canvas.draw_circle((cx, cy), radius, &paint);
        });
    }

    // ========== LocalImageCapable Methods ==========

    /// Draw an image from a local file.
    pub fn draw_image(
        &mut self,
        file_path: &str,
        x: f32,
        y: f32,
        width: f32,
        height: f32,
        use_cache: bool,
    ) -> Result<()> {
        let image = load_image(file_path, use_cache)?;
        let dest_rect = Rect::from_xywh(x, y, width, height);

        self.with_canvas(|canvas| {
            canvas.draw_image_rect(&image, None, dest_rect, &Paint::default());
        });

        Ok(())
    }

    /// Measure the size of an image from a local file.
    pub fn measure_image(&self, file_path: &str, use_cache: bool) -> Result<(i32, i32)> {
        measure_image(file_path, use_cache)
    }
}

// Internal helper methods
impl<'a> SkiaPainter<'a> {
    /// Access the canvas and run a closure.
    fn with_canvas<F>(&mut self, f: F)
    where
        F: FnOnce(&Canvas),
    {
        let canvas = self.surface.canvas();
        f(canvas);
    }

    /// Create a Paint for fill operations.
    fn make_fill_paint(&self) -> Paint {
        let mut paint = Paint::default();
        paint.set_style(PaintStyle::Fill);
        paint.set_anti_alias(true);

        if let Some(ref color_str) = self.current_style.fill_color {
            paint.set_color(parse_color(color_str));
        }

        paint
    }

    /// Create a Paint for stroke operations.
    fn make_stroke_paint(&self) -> Paint {
        let mut paint = Paint::default();
        paint.set_style(PaintStyle::Stroke);
        paint.set_anti_alias(true);
        paint.set_stroke_width(self.current_style.stroke_width);

        if let Some(ref color_str) = self.current_style.stroke_color {
            paint.set_color(parse_color(color_str));
        } else if let Some(ref color_str) = self.current_style.fill_color {
            // Fall back to fill color if no stroke color
            paint.set_color(parse_color(color_str));
        }

        paint
    }

    /// Internal text drawing with font fallback support using Paragraph.
    fn draw_text_internal(
        &mut self,
        text: &str,
        x: f32,
        y: f32,
        _max_width: Option<f32>,
        _paint: &Paint,
    ) {
        if text.is_empty() {
            return;
        }

        let font_collection = get_font_collection();
        let paragraph_style = ParagraphStyle::new();

        let mut text_style = TextStyle::new();
        text_style.set_font_size(self.current_font.size());

        // Set font family
        if let Some(ref family) = self.current_style.font_family {
            text_style.set_font_families(&[family.as_str()]);
        }

        // Set color from current style
        if let Some(ref color_str) = self.current_style.fill_color {
            text_style.set_color(parse_color(color_str));
        } else {
            text_style.set_color(Color::BLACK);
        }

        let mut builder = ParagraphBuilder::new(&paragraph_style, font_collection);
        builder.push_style(&text_style);
        builder.add_text(text);

        let mut paragraph = builder.build();
        paragraph.layout(f32::MAX);

        // Paragraph draws from top-left, but fill_text expects baseline position
        // Use the paragraph's own alphabetic baseline for accurate positioning
        let baseline = paragraph.alphabetic_baseline();

        self.with_canvas(|canvas| {
            paragraph.paint(canvas, (x, y - baseline));
        });
    }
}
